### ISSUE TITLE
ZSH completion does not allow to specify multiple files to trash

### ISSUE THREAD

**Issue by ohquero** [andreafrancia/trash-cli#338]
**Describe the bug**
Commands provided by this utility allow to specify multiple files to manipulate, but ZSH autocompletions suggest filenames only for one parameter. 

**trash-cli version**
0.23.11.10

**Are you using the latest version of trash-cli?**
Yes

**Have you tried if the bug is present in the latest version of trash-cli?**
Yes

**Operating system:**
Arch Linux

**To Reproduce**
```
$ touch bar foo
$ trash-put bar f#Here ZSH autocompletion does not complete with "foo" file
$ trash-put bar foo #But the command works!
$ trash-list
2024-03-28 09:57:28 /home/quero/bar
2024-03-28 09:57:28 /home/quero/foo
```

**Expected behavior**
The completions should have completed the command the "foo" filename.

**Comment by DMaroo** [comment]
## Patch

The attached patch, when applied to the generated completions, fixes the issue.

```diff
23c23
<   "(*)::files:_trash_files"
---
>   "*:: :->file"
28c28
<   local context state line curcontext="$curcontext" one_or_more='(-)*' remainder='(*)'
---
>   local context state line curcontext="$curcontext"
30,33c30
<   if ((${_shtab_trash_put_options[(I)${(q)one_or_more}*]} + ${_shtab_trash_put_options[(I)${(q)remainder}*]} == 0)); then  # noqa: E501
<     _shtab_trash_put_options+=(': :_shtab_trash_put_commands' '*::: :->trash-put')
<   fi
<   _arguments -C -s $_shtab_trash_put_options
---
>   _arguments -C -s -S -A -* $_shtab_trash_put_options
42a40,45
>       ;;
>     file)
>       (( CURRENT > 0 )) && line[CURRENT]=()
>       line=( ${line//(#m)[\[\]()\\*?#<>~\^\|]/\\$MATCH} )
>       _files -F line && ret=0
>       ;;
```

Sorry, I can't attach the patch file because GitHub [does not let](https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/attaching-files) Linux users upload `.patch` files. However, I am attaching the patch as a `.txt` file: [trash-cli_zsh-completion.txt](https://github.com/andreafrancia/trash-cli/files/14904531/trash-cli_zsh-completion.txt).

Of course, this is a hacky fix and it should _not_ be upstreamed. I believe that either there is something wrong in the way [`shtab`](https://github.com/iterative/shtab) is used to generate completions, or `shtab` has a bug (unlikely).

## Behavior

When using `-<TAB>` without files as any previous arguments, all the flags get auto-completed.

```
$ trash -<TAB>
--force             -f                       -- silently ignore nonexistent files
--help                                       -- show this help message and exit
--interactive       -i                       -- prompt before every removal
--print-completion                           -- print shell completion script
--recursive         --directory  -R  -r  -d  -- ignored (for GNU rm compatibility)
--trash-dir                                  -- use TRASHDIR as trash folder
--verbose           -v                       -- explain what is being done
--version                                    -- show program's version number and exit
```

When using just `<TAB>` (with no `-` in front of it) with or without files as any previous arguments, the files get autocompleted. If you have a file as one of the previous arguments, and then you use `-<TAB>`, it will try autocompleting files and will not autocomplete flags.

This behavior is completely identical to `rm` behavior.

**Comment by protist** [comment]
@DMaroo just wondering why this shouldn't be upstreamed? Just curious as to whether users should patch this downstream.

**Comment by DrummyFloyd** [comment]
any update on this ?

**Comment by DMaroo** [comment]
> [@DMaroo](https://github.com/DMaroo) just wondering why this shouldn't be upstreamed? Just curious as to whether users should patch this downstream.

Ideally, `shtab` should have generated that code, but it did not. Patching the auto-generated file is not the right way to go IMO.

**Comment by protist** [comment]
Thanks @DMaroo . I had a look in my system, and I can't find the file to patch. Could you please explain what we should be patching?

**Comment by dnswrsrx** [comment]
@protist it's one of the directories where ZSH stores completions. You may have to look it up for the OS you're running.

I run Arch, and the completions for `trash` are stored in `/usr/share/zsh/site-functions`. In there, the completion for the `trash` command is named `_trash`, so I'm applying said patch to this file. Again with my OS, my command to patch the file is `patch _trash <patch file name>`.

I had to update the patch for it to work on my system. Here's my [gist](https://gist.github.com/dnswrsrx/24d264d5e449ea1d96903d35919dcf1b) in case it also helps you.

**Comment by protist** [comment]
Thanks @dnswrsrx. That works great on my system (also Arch). I'm considering opening a task on the Arch tracker to patch it there, since there hasn't been much movement here.

**Comment by andreafrancia** [comment]
I tried on my computer the newest version of trash-cli but I cannot reproduce the failure.
Maybe in between was solved in some commit.
I'm closing this issue, fell free to comment below of reopen another issue if the problem persist.

**Comment by protist** [comment]
I also tried to replicate this bug on an un-patched system, and it does seem to work now! Good outcome; thank you @andreafrancia!

**Comment by revsuine** [comment]
Interesting. On Artix with trash-cli 0.24.5.26-2, the issue still persists. I wonder if it's a packaging issue

**Comment by protist** [comment]
Ech, I just tested on another Arch system, and it's still broken there. I'm not sure what I did with the working system.

E: I can confirm if I apply the previous [patch](https://github.com/andreafrancia/trash-cli/issues/338#issuecomment-2042457689) to [`/usr/share/zsh/site-functions/_trash`](https://github.com/andreafrancia/trash-cli/issues/338#issuecomment-2615039155) it still fixes the issue.

**Comment by JPFrancoia** [comment]
Hey, the patch doesn't fully apply for trash-cli 0.24.5.26-3 (on Arch):

```shell
patching file _trash
Hunk #2 FAILED at 28.
Hunk #3 FAILED at 30.
2 out of 4 hunks FAILED -- saving rejects to file _trash.rej
```

rejection file:
```
*** /dev/null
--- /dev/null
***************
*** 28
-   local context state line curcontext="$curcontext" one_or_more='(-)*' remainder='(*)'
--- 28 -----
+   local context state line curcontext="$curcontext"
***************
*** 30,33
-   if ((${_shtab_trash_put_options[(I)${(q)one_or_more}*]} + ${_shtab_trash_put_options[(I)${(q)remainder}*]} == 0)); then  # noqa: E501
-     _shtab_trash_put_options+=(': :_shtab_trash_put_commands' '*::: :->trash-put')
-   fi
-   _arguments -C -s $_shtab_trash_put_options
--- 30 -----
+   _arguments -C -s -S -A -* $_shtab_trash_put_options
```

**Comment by protist** [comment]
I can confirm that the patches no longer work for me in the same version in Arch. I tried @DMaroo 's [original patch](https://github.com/andreafrancia/trash-cli/issues/338#issuecomment-2042457689) and @dnswrsrx 's [patch](https://github.com/andreafrancia/trash-cli/issues/338#issuecomment-2615039155) that was [updated](https://gist.github.com/dnswrsrx/24d264d5e449ea1d96903d35919dcf1b) 2 weeks ago. I get the same output for the first patch, but even for the second newer patch, I still get

```
$ patch /tmp/_trash < /tmp/_trash.patch
patching file /tmp/_trash
Hunk #1 FAILED at 20.
patch unexpectedly ends in middle of line
Hunk #2 succeeded at 48 with fuzz 2 (offset 1 line).
1 out of 2 hunks FAILED -- saving rejects to file /tmp/_trash.rej
```

with rejection file
```
--- /dev/null
+++ /dev/null
@@ -20,23 +20,12 @@
   "--trash-dir[use TRASHDIR as trash folder]:trashdir:(\${\$(trash-list --trash-dirs)#parent_*})"
   "*"{-v,--verbose}"[explain what is being done]"
   "(- : *)--version[show program\'s version number and exit]"
-  "(*)::files:_trash_files"
+  "*:: :->file"
 )
 
-# guard to ensure default positional specs are added only once per session
-_shtab_trash_defaults_added=0
-
-
 _shtab_trash() {
-  local context state line curcontext="$curcontext" one_or_more='(*)' remainder='(-)*' default='*::: :->trash'
+  local context state line curcontext="$curcontext"
 
-  # Add default positional/remainder specs only if none exist, and only once per session
-  if (( ! _shtab_trash_defaults_added )); then
-    if (( ${_shtab_trash_options[(I)${(q)one_or_more}*]} +          ${_shtab_trash_options[(I)${(q)remainder}*]} +          ${_shtab_trash_options[(I)${(q)default}]} == 0 )); then
-      _shtab_trash_options+=(': :_shtab_trash_commands' '*::: :->trash')
-    fi
-    _shtab_trash_defaults_added=1
-  fi
   _arguments -C -s $_shtab_trash_options
 
   case $state in
```

Also, in case it was unclear, @andreafrancia could you please re-open this issue?

**Comment by Inventhrice** [comment]
@protist @JPFrancoia 

Just a heads up, this [response](https://github.com/andreafrancia/trash-cli/issues/363#issuecomment-2652603487) does work for me!
```
compdef _gnu_generic trash
````

**Comment by protist** [comment]
Thanks @Inventhrice ! That seem to work well for me too!

**Comment by JPFrancoia** [comment]
Works for me too 🚀 . But don't forget to delete any leftover completion files:

```
cd /usr/share/zsh/site-functions/
sudo rm *trash*
```
